Update Documentation: Optimized Schemes, PERK Section, and PairedExplicitRK2 Tutorial (#2146)
trixi-framework/Trixi.jl: docs/src/time_integration.md

Formally, this boils down to an optimization problem of the form 
```math
\underset{P_{p;S} \, \in \, \mathcal{P}_{p;S}}{\max} \Delta t \text{ such that } \big \vert P_{p;S}(\Delta t \lambda_m) \big \vert \leq 1, \quad  m = 1 , \dots , M \tag{1}
```
where $p$ denotes the order of consistency of the scheme, $S$ is the number of stage evaluations and $M$ denotes the number of eigenvalues $\lambda_m$ of the Jacobian matrix $J \coloneqq \frac{\partial \boldsymbol F}{\partial \boldsymbol U}$ of the right-hand side of the [semidiscretized PDE](https://trixi-framework.github.io/Trixi.jl/stable/overview/#overview-semidiscretizations):
```math
\dot{\boldsymbol U} = \boldsymbol F(\boldsymbol U) \tag{2} \: .
```
In particular, for $S > p$ the Runge-Kutta method includes some free coefficients which may be used to adapt the domain of absolute stability to the problem at hand.
Since Trixi.jl [supports exact computation of the Jacobian $J$ by means of automatic differentiation](https://trixi-framework.github.io/Trixi.jl/stable/tutorials/differentiable_programming/), we have access to the Jacobian of a given simulation setup.

> For small (up to approximately $10^4$ DoF) systems, the spectrum $\boldsymbol \sigma = \left \{ \lambda_m \right \}_{m=1, \dots, M}$ can be computed directly using [`LinearAlgebra.eigvals(J)`](https://docs.julialang.org/en/v1/stdlib/LinearAlgebra/#LinearAlgebra.eigvals).

Reviewer: ```suggestion
For small (say, up to roughly $10^4$ DoF) systems, the spectrum $\boldsymbol \sigma = \left \{ \lambda_m \right \}_{m=1, \dots, M}$ can be computed directly using [`LinearAlgebra.eigvals(J)`](https://docs.julialang.org/en/v1/stdlib/LinearAlgebra/#LinearAlgebra.eigvals).
```

The optimization problem (1) can be solved using the algorithms described in [Ketcheson, Ahmadia (2012)](http://dx.doi.org/10.2140/camcos.2012.7.247) for a moderate number of stages $S$ or [Doehring, Gassner, Torrilhon (2024)](https://doi.org/10.1007/s10915-024-02478-5) for a large number of stages $S$.
In Trixi.jl, the former approach is implemented by means of convex optimization using the [Convex.jl](https://github.com/jump-dev/Convex.jl) package.

The resulting stability polynomial $P_{p;S}$ is then used to construct an optimized Runge-Kutta method.
Trixi.jl implements the [Paired-Explicit Runge-Kutta (PERK)](https://doi.org/10.1016/j.jcp.2019.05.014) method, a low-storage, multirate-ready method with optimized domain of absolute stability.

### Paired-Explicit Runge-Kutta (PERK) Schemes

Paired-Explicit Runge-Kutta (PERK) or `PairedExplicitRK` schemes are an advanced class of numerical methods designed to efficiently solve ODEs.

> In the [original publication](https://doi.org/10.1016/j.jcp.2019.05.014), second-order schemes were introduced, which have been extended to [third](https://doi.org/10.1016/j.jcp.2022.111470)- and [fourth](https://doi.org/10.48550/arXiv.2408.05470)-order in subsequent works.

Reviewer: ```suggestion
In the [original publication](https://doi.org/10.1016/j.jcp.2019.05.014), second-order schemes were introduced, which have been extended to [third](https://doi.org/10.1016/j.jcp.2022.111470) and [fourth](https://doi.org/10.48550/arXiv.2408.05470) order in subsequent works.
```

By construction, PERK schemes are suited for integrating multirate systems, i.e., systems with varying characteristics speeds throughout the domain.

> Nevertheless, due to their optimized stability properties and low-storage nature, the PERK schemes are also highly efficient when applied as standalone methods. In Trixi.jl, the standalone PERK integrators require that all stages of the method be active.

Reviewer: ```suggestion
Nevertheless, due to their optimized stability properties and low-storage nature, the PERK schemes are also highly efficient when applied as standalone methods. In Trixi.jl, the _standalone_ PERK integrators require that all stages of the method are active.
```

#### Tutorial: Using `PairedExplicitRK2`

In this tutorial, we will demonstrate how you can use the second-order PERK time integrator. You need the packages `Convex.jl` and `ECOS.jl`, so be sure they are added to your environment.

1. First, you need to load the necessary packages:

> ```@example 1

Reviewer 2: ````suggestion
```@example PERK-example-1
````

The names are global

Author: I see!

using Convex, ECOS
using OrdinaryDiffEq
using Trixi
```

2. Define the ODE problem and the semidiscretization setup. For this example, we will use a simple advection problem.

> ```@example 1

Reviewer 2: ````suggestion
```@example PERK-example-1
````

# Define the mesh
cells_per_dimension = 100
coordinates_min = 0.0
coordinates_max = 1.0
mesh = StructuredMesh(cells_per_dimension, 
                      coordinates_min, coordinates_max)

# Define the equation and initial condition
advection_velocity = 1.0
equations = LinearScalarAdvectionEquation1D(advection_velocity)

initial_condition = initial_condition_convergence_test

# Define the solver
solver = DGSEM(polydeg = 3, surface_flux = flux_lax_friedrichs)

# Define the semidiscretization
semi = SemidiscretizationHyperbolic(mesh, 
                                    equations, initial_condition, 
                                    solver)
```

> 3. Define the necessary callbacks for the simulation. Callbacks are used to perform actions at specific points during the integration process.

Reviewer 2: Can we link to the callbacks page of the docs, please?

> ```@example 1

Reviewer 2: ````suggestion
```@example PERK-example-1
````

> 3. Define the necessary [callbacks](callbacks.md) for the simulation. Callbacks are used to perform actions at specific points during the integration process.

Reviewer 3: ```suggestion
3. Define the necessary [callbacks](@ref callbacks-id) for the simulation. Callbacks are used to perform actions at specific points during the integration process.
```
Does the link work as you intend? I would have expected something using `@ref`, which uses the [identifier](https://github.com/trixi-framework/Trixi.jl/blob/897c1cdf693321aa4b52e9737381661223d65645/docs/src/callbacks.md#L1C19-L1C31) of the caption.

Author: It does, but your suggestion is more uniform to the way the rest of the things in the documentation is linked. Thx!

Reviewer 3: Interesting that linking to the markdown file itself also works. I didn't know that.

# Define some standard callbacks
summary_callback  = SummaryCallback()
alive_callback    = AliveCallback()
analysis_callback = AnalysisCallback(semi, interval = 200)
# For this optimized method we can use a relatively large CFL number
stepsize_callback = StepsizeCallback(cfl = 2.5)

# Create a CallbackSet to collect all callbacks
callbacks = CallbackSet(summary_callback, 
                        alive_callback, 
                        analysis_callback, 
                        stepsize_callback)
```

4. Define the ODE problem by specifying the time span over which the ODE will be solved. 
The `tspan` parameter is a tuple `(t_start, t_end)` that defines the start and end times for the simulation. 
The `semidiscretize` function is used to create the ODE problem from the simulation setup.

> ```@example 1

Reviewer 2: ````suggestion
```@example PERK-example-1
````

# Define the time span
tspan = (0.0, 1.0)

# Create ODE problem with time span from 0.0 to 1.0
ode = semidiscretize(semi, tspan)
```

5. In this step we will construct the time integrator. In order to do this, you need the following components:

>   - Number of Stages: The number of stages $S$ in the Runge-Kutta method. 

Reviewer 2: ```suggestion
  - Number of stages: The number of stages $S$ in the Runge-Kutta method. 
```

  In this example, we use `6` stages.

>   - Time Span (`tspan`): A tuple `(t_start, t_end)` that defines the time span over which the ODE will be solved. 

Reviewer 2: ```suggestion
  - Time span (`tspan`): A tuple `(t_start, t_end)` that defines the time span over which the ODE will be solved. 
```

  This defines the bounds for the bisection routine for the optimal timestep $\Delta t$ used in calculating the polynomial coefficients at optimization stage. 
  This variable is already defined in step 5.
  - Semidiscretization (`semi`): The semidiscretization setup that includes the mesh, equations, initial condition, and solver. In this example, this variable is already defined in step 3.
  In the background, we compute from `semi` the Jacobian $J$ evaluated at the initial condition using [`jacobian_ad_forward`](https://trixi-framework.github.io/Trixi.jl/stable/reference-trixi/#Trixi.jacobian_ad_forward-Tuple{Trixi.AbstractSemidiscretization}).
  This is then followed by the computation of the spectrum $\boldsymbol \sigma(J)$ using `LinearAlgebra.eigvals`.
  Equipped with the spectrum, the optimal stability polynomial is computed, from which the corresponding Runge-Kutta method is derived. Other constructors (if the coefficients $\boldsymbol{\alpha}$ of the stability polynomial are already available, or if a reduced spectrum $\widetilde{\boldsymbol{\sigma}}$ should be used) are discussed below.

> ```@example 1

Reviewer 2: ````suggestion
```@example PERK-example-1
````

# Construct second order-explicit Runge-Kutta method with 6 stages for given simulation setup (`semi`)
# `tspan` provides the bounds for the bisection routine that is used to calculate the maximum timestep.
ode_algorithm = Trixi.PairedExplicitRK2(6, tspan, semi)
```

6. With everything now set up, you can now use `Trixi.solve` to solve the ODE problem. The `solve` function takes the ODE problem, the time integrator, and some options such as the time step (`dt`), whether to save every step (`save_everystep`), and the callbacks.

> ```@example 1

Reviewer 2: ````suggestion
```@example PERK-example-1
````

# Solve the ODE problem using PERK2
sol = Trixi.solve(ode, ode_algorithm,
                  dt = 1.0, # overwritten by `stepsize_callback`
                  save_everystep = false, callback = callbacks)
```

##### Advanced constructors:
There are two additional constructors for the `PairedExplicitRK2` method besides the one taking in a semidiscretization `semi`:
  - `PairedExplicitRK2(num_stages, base_path_monomial_coeffs::AbstractString)` constructs a `num_stages`-stage method from the given optimal monomial_coeffs $\boldsymbol \alpha$.
  These are expected to be present in the provided directory in the form of a `gamma_<S>.txt` file, where `<S>` is the number of stages `num_stages`.
  This constructor is useful when the optimal coefficients cannot be obtained using the optimization routine by Ketcheson and Ahmadia, possibly due to a large number of stages $S$.
  - `PairedExplicitRK2(num_stages, tspan, eig_vals::Vector{ComplexF64})` constructs a `num_stages`-stage using the optimization approach by Ketcheson and Ahmadia for the (reduced) spectrum `eig_vals`.
  The use-case for this constructor would be a large system, for which the computation of all eigenvalues is infeasible.

> #### Automatic computation of stable CFL Number

Reviewer: ```suggestion
#### Automatic computation of a stable CFL Number
```

In the previous tutorial the CFL number was set manually to $2.5$.

> To avoid this trial-and error process, instantiations of `AbstractPairedExplicitRK` methods can automatically compute the stable CFL number for a given simulation setup using the [`Trixi.calculate_cfl`](@ref) function.

Reviewer: ```suggestion
To avoid the manual trial-and-error process behind this, instantiations of `AbstractPairedExplicitRK` methods can automatically compute the stable CFL number for a given simulation setup using the [`Trixi.calculate_cfl`](@ref) function.
```

When constructing the time integrator from a semidiscretization `semi`, 
```@example 1
# Construct third-order paired-explicit Runge-Kutta method with 8 stages for given simulation setup.                               
ode_algorithm = Trixi.PairedExplicitRK3(8, tspan, semi)
```
the maximum timestep `dt` is stored by the `ode_algorithm`.
This can then be used to compute the stable CFL number for the given simulation setup:

> ```@example 1

Reviewer 2: ````suggestion
```@example PERK-example-1
````

cfl_number = Trixi.calculate_cfl(ode_algorithm, ode)
```
For nonlinear problems, the spectrum will in general change over the course of the simulation.
Thus, it is often necessary to reduce the optimal `cfl_number` by a safety factor:

> ```@example 1

Reviewer 2: ````suggestion
```@example PERK-example-1
````

# For non-linear problems, the CFL number should be reduced by a safety factor
# as the spectrum changes (in general) over the course of a simulation
stepsize_callback = StepsizeCallback(cfl = 0.85 * cfl_number)
```
If the optimal monomial coefficients are precomputed, the user needs to set the obtained maximum timestep from the optimization manually via

> ```@example 1

Reviewer 2: ````suggestion
```@example PERK-example-1
````

In the [original publication](https://doi.org/10.1016/j.jcp.2019.05.014), second-order schemes were introduced, which have been extended to [third](https://doi.org/10.1016/j.jcp.2022.111470)- and [fourth](https://doi.org/10.48550/arXiv.2408.05470)-order in subsequent works.

By construction, P-ERK schemes are suited for integrating multirate systems, i.e., systems with varying characteristics speeds throughout the domain.
Nevertheless, due to their optimized stability properties and low-storage nature, the P-ERK schemes are also highly efficient when applied as standalone methods.

#### Tutorial: Using `PairedExplicitRK2`

In this tutorial, we will demonstrate how you can use the second-order P-ERK time integrator.

1. First, ensure you have the necessary packages installed. For the optimization of the stability polynomial $P_{2; S}$, you need the `Convex.jl` and `ECOS.jl` packages.
You can install them using Julia's package manager:

```julia
using Pkg
Pkg.add("Trixi")
Pkg.add("Convex")
Pkg.add("ECOS")
```

2. Now you can load the necessary packages:

```julia
using Convex, ECOS
using Trixi

> ```

Reviewer 2: Can you please turn the code snippets into running code? Otherwise, it will get out of date and will become useless over time.

Reviewer 3: I guess you could use `@example` (see [documentation](https://documenter.juliadocs.org/stable/man/syntax/#reference-at-example)).

Author: I have modified some parts to use `@setup`, `@repl`, and `@example` instead. Now when I build the documentation from (at least on my local PC) it added `Trixi`, `Convex`, and `ECOS` to `Project.toml`. I am not sure how to stop this from happening. Is this OK?

Reviewer: So you say `Trixi` is added to the `Trixi.jl` `Project.toml` ? That sounds indeed not desirable.

Author: The main configuration file of the package is intact but `Trixi.jl` is added to `Project.toml` in `docs`.

Reviewer 3: Adding Convex and ECOS to `docs/Project.toml` is fine. Please also provide appropriate compat bounds (e.g. take the ones from the main `Project.toml`). But you should not explicitly `Pkg.add` them (and also not Trixi.jl) in the docs pages. If they are in `docs/Project.toml` that is enough. Trixi.jl should point to the development version in `..` from `docs`, i.e. it will be added once you do `julia --project=docs -e 'using Pkg; Pkg.develop(PackageSpec(path=pwd())); Pkg.instantiate()''` locally (or [in CI](https://github.com/trixi-framework/Trixi.jl/blob/3609f39b979de2e894bb78ecb9c67e8fe75ddf9f/.github/workflows/Documenter.yml#L46)). If you have Trixi.jl in `docs/Project.toml` locally you can just not push that.

Author: I see! I reckon the intention of this code block isn’t mainly to add the packages `Convex`, `ECOS`, and `Trixi` so the rest of the codes in the docs pages run fine but also to guide the user that these installations are necessary to use the optimized scheme. But if you don’t think it’s a good idea to add the packages in the documentation pages, I can also simply try to change these instructions to only be some texts instead.

Reviewer 3: I think it's not necessary to explicitly add the packages in the documentation pages. We also don't do that for the other tutorials (regarding, e.g., OrdinaryDiffEq.jl or Plots.jl). You can, however, emphasize that one needs to add these packages. I'll make a suggestion.

Author: Got it! Thanks for the suggestion!

The resulting stability polynomial $P_{p;S}$ is then used to construct an optimized Runge-Kutta method.
Trixi.jl implements the [Paired-Explicit Runge-Kutta (P-ERK)](https://doi.org/10.1016/j.jcp.2019.05.014) method, a low-storage, multirate-ready method with optimized domain of absolute stability.

### Paired-Explicit Runge-Kutta (P-ERK) Schemes

Paired-Explicit Runge-Kutta (P-ERK) or `PairedExplicitRK` schemes are an advanced class of numerical methods designed to efficiently solve ODEs.
In the [original publication](https://doi.org/10.1016/j.jcp.2019.05.014), second-order schemes were introduced, which have been extended to [third](https://doi.org/10.1016/j.jcp.2022.111470)- and [fourth](https://doi.org/10.48550/arXiv.2408.05470)-order in subsequent works.

By construction, P-ERK schemes are suited for integrating multirate systems, i.e., systems with varying characteristics speeds throughout the domain.
Nevertheless, due to their optimized stability properties and low-storage nature, the P-ERK schemes are also highly efficient when applied as standalone methods.

#### Tutorial: Using `PairedExplicitRK2`

> In this tutorial, we will demonstrate how you can use the second-order P-ERK time integrator.
> 
> 1. First, ensure you have the necessary packages installed. For the optimization of the stability polynomial $P_{2; S}$, you need the `Convex.jl` and `ECOS.jl` packages.
> You can install them using Julia's package manager:
> 
> ```@repl 1
> using Pkg
> Pkg.add("Trixi")
> Pkg.add("Convex")
> Pkg.add("ECOS")

Reviewer 3: ```suggestion
In this tutorial, we will demonstrate how you can use the second-order P-ERK time integrator. You need the packages `Convex.jl` and `ECOS.jl`, so be sure they are added to your environment.
```
Maybe something like this?

```

> 2. Now you can load the necessary packages:

Reviewer 3: ```suggestion
1. First, you need to load the necessary packages:
```

```@setup 1
using Convex, ECOS
using OrdinaryDiffEq
using Trixi
```

> 3. Define the ODE problem and the semidiscretization setup. For this example, we will use a simple advection problem.

Reviewer 3: ```suggestion
2. Define the ODE problem and the semidiscretization setup. For this example, we will use a simple advection problem.
```
See above.

```@example 1
# Define the mesh
cells_per_dimension = 100
coordinates_min = 0.0
coordinates_max = 1.0
mesh = StructuredMesh(cells_per_dimension, 
                      coordinates_min, coordinates_max)

# Define the equation and initial condition
advection_velocity = 1.0
equations = LinearScalarAdvectionEquation1D(advection_velocity)

initial_condition = initial_condition_convergence_test

# Define the solver
solver = DGSEM(polydeg = 3, surface_flux = flux_lax_friedrichs)

# Define the semidiscretization
semi = SemidiscretizationHyperbolic(mesh, 
                                    equations, initial_condition, 
                                    solver)
```

> 4. Define the necessary callbacks for the simulation. Callbacks are used to perform actions at specific points during the integration process.

Reviewer 3: ```suggestion
3. Define the necessary callbacks for the simulation. Callbacks are used to perform actions at specific points during the integration process.
```
See above.

```@example 1
# Define some standard callbacks
summary_callback  = SummaryCallback()
alive_callback    = AliveCallback()
analysis_callback = AnalysisCallback(semi, interval = 200)
# For this optimized method we can use a relatively large CFL number
stepsize_callback = StepsizeCallback(cfl = 2.5)

# Create a CallbackSet to collect all callbacks
callbacks = CallbackSet(summary_callback, 
                        alive_callback, 
                        analysis_callback, 
                        stepsize_callback)
```

> 5. Define the ODE problem by specifying the time span over which the ODE will be solved. 

Reviewer 3: ```suggestion
4. Define the ODE problem by specifying the time span over which the ODE will be solved. 
```
See above.

The `tspan` parameter is a tuple `(t_start, t_end)` that defines the start and end times for the simulation. 
The `semidiscretize` function is used to create the ODE problem from the simulation setup.

```@example 1
# Define the time span
tspan = (0.0, 1.0)

# Create ODE problem with time span from 0.0 to 1.0
ode = semidiscretize(semi, tspan)
```

> 6. In this step we will construct the time integrator. In order to do this, you need the following components:

Reviewer 3: ```suggestion
5. In this step we will construct the time integrator. In order to do this, you need the following components:
```

  - Semidiscretization (`semi`): The semidiscretization setup that includes the mesh, equations, initial condition, and solver. In this example, this variable is already defined in step 3.
  In the background, we compute from `semi` the Jacobian $J$ evaluated at the initial condition using [`jacobian_ad_forward`](https://trixi-framework.github.io/Trixi.jl/stable/reference-trixi/#Trixi.jacobian_ad_forward-Tuple{Trixi.AbstractSemidiscretization}).
  This is then followed by the computation of the spectrum $\boldsymbol \sigma(J)$ using `LinearAlgebra.eigvals`.
  Equipped with the spectrum, the optimal stability polynomial is computed, from which the corresponding Runge-Kutta method is derived. Other constructors (if the coefficients $\boldsymbol{\alpha}$ of the stability polynomial are already available, or if a reduced spectrum $\widetilde{\boldsymbol{\sigma}}$ should be used) are discussed below.

```@example 1
# Construct second order-explicit Runge-Kutta method with 6 stages for given simulation setup (`semi`)
# `tspan` provides the bounds for the bisection routine that is used to calculate the maximum timestep.
ode_algorithm = Trixi.PairedExplicitRK2(6, tspan, semi)
```

> 7. With everything now set up, you can now use `Trixi.solve` to solve the ODE problem. The `solve` function takes the ODE problem, the time integrator, and some options such as the time step (`dt`), whether to save every step (`save_everystep`), and the callbacks.

Reviewer 3: ```suggestion
6. With everything now set up, you can now use `Trixi.solve` to solve the ODE problem. The `solve` function takes the ODE problem, the time integrator, and some options such as the time step (`dt`), whether to save every step (`save_everystep`), and the callbacks.
```

```@example 1
# Solve the ODE problem using PERK2
sol = Trixi.solve(ode, ode_algorithm,
                  dt = 1.0, # overwritten by `stepsize_callback`
                  save_everystep = false, callback = callbacks)
```

> 8. Advanced constructors:

Reviewer 3: ```suggestion
7. Advanced constructors:
```
